### PR TITLE
Refactors ALL THE GLOBAL THINGS.

### LINKED ISSUE #5497: non-admins can't access any campaign pages because of double-country code in URL
As a user coming from Mexico, if I filter home page campaigns by "Environment" I see results which include the following:

image

When selecting the "Fish Bowl Baggies" campaign, I get a 404 page and the URL has duplicate country prefixes which is likely the cause for the 404:

`https://thor.dosomething.org/mx/mx/voluntario/fish-bowl-baggies`

image

### LINKED ISSUE #5506: /user links are pre-appended with /us
when I login from 
`https://thor.dosomething.org/mx` I am logged in and then get redirected to `https://thor.dosomething.org/us/user/3118670`
it should be
`https://thor.dosomething.org/mx/user/3118670`

@mshmsh5000 also noticed password redirect links are getting sent to `us/user/password-reset`

### PR DESCRIPTION
**Pull request by itsjoekent**
ALL OF THEM.
#### What's this PR do?
- Refactors the dosomething_global_init() to be closer to english than brainfuck
  - Gathers all variables needed at the start using helper functions
  - Breaks conditional logic out into new function and cuts duplicated, confusing or un-needed code
- Adds dosomething_global_is_path_prefix to check if a string is a path prefix
- Condenses the node redirect logic into one small function `dosomething_global_redirect_node` (thanks to the refactoring from above)

And yes, there are lots of comments.
#### Where should the reviewer start?

init()
#### How should this be manually tested?

Where was that doc with all of the test cases again? We'll be needed that...
#### Any background context you want to provide?

See my or andrea's brain.

TLDR: The old global init() was a mess from all of us adding in new URL edge cases and wasn't maintainable, causing multiple bugs.
#### What are the relevant tickets?

Fixes #5506
Fixes #5497

### CODE REVIEW

File: lib/modules/dosomething/dosomething_global/dosomething_global.module [hunk 1]
```diff
@@ -93,36 +76,91 @@ function dosomething_global_init() {
...
-  }
+/**
+ * Extract relevant user variables from the current request
+ * to be used in dosomething_global_init().
+ *
+ * @return array
+ *      header_country_code - Country code specified in the headers
+ *      user_language - Matching language, or NULL for no matching installed language
+ *      user_path_prefix - The prefix which matches the users language
+ *      is_user_page - Is this the user profile page?
```
Comment (angaither) [inline lib/modules/dosomething/dosomething_global/dosomething_global.module]: it would be helpful to know types of these vars

File: lib/modules/dosomething/dosomething_global/dosomething_global.module [hunk 2]
```diff
@@ -413,146 +461,76 @@ function dosomething_global_get_language($account, stdClass $node = NULL) {
...
-  $target_lang_code = DOSOMETHING_GLOBAL_DEFAULT_LANG_CODE;
-
-  $lang_map = variable_get('dosomething_global_language_map', '');
-  foreach ($lang_map as $lang_key => $lang) {
-    if ($lang['country'] == $user_country_code) {
-      $target_lang_code = $lang_key;
-      break;
+  // Don't do redirects for privileged users.
+  if ($redirect_status == 302) {
+    if (user_access('access administration menu')) {
```
Comment (angaither) [inline lib/modules/dosomething/dosomething_global/dosomething_global.module:506]: I don't see this in the previous version, this means anyone on our staff (admin, editor, site admin, etc) and regional admins won't get redirected here.. is that intended?

Comment (itsjoekent) [inline lib/modules/dosomething/dosomething_global/dosomething_global.module:506]: https://github.com/DoSomething/phoenix/blob/global/lib/modules/dosomething/dosomething_global/dosomething_global.module#L457

File: lib/modules/dosomething/dosomething_global/dosomething_global.module [hunk 3]
```diff
@@ -14,77 +14,60 @@ define('DOSOMETHING_GLOBAL_DEFAULT_LANG_CODE', 'en-global');
...
-
-      // Append the user language
-      $lang_append = $new_translation_prefix . $user_language;
-
-      // Build complete URL and redirect
-      $new_url = str_replace(' ', '', '/node/' . $nid . '/edit/' . $lang_append);
-      drupal_goto($new_url);
-    }
+  // Configure the correct path prefix
+  $url_variables['url_path_prefix'] = dosomething_global_get_prefix_for_language($user_variables['user_language']);
```
Comment (angaither) [inline lib/modules/dosomething/dosomething_global/dosomething_global.module:50]: can this prefix stuff be refactored a bit to use the function @sbsmith86 just wrote? `dosomething_global_get_current_prefix()`

File: lib/modules/dosomething/dosomething_global/dosomething_global.module (hunk 2 above)
Comment (angaither) [inline lib/modules/dosomething/dosomething_global/dosomething_global.module:506]: ahhh, hidden way above :+1:

**Comment by angaither** [comment]
:scream: let's try this :+1:

**Comment by itsjoekent** [comment]
:boom:
